### PR TITLE
Only notify ready global checkpoint listeners

### PR DESCRIPTION
**Pull request by jasontedor**
When we add a global checkpoint listener, it is also carries along with it a value that it thinks is the current global checkpoint. This value can be above the actual global checkpoint on a shard if the listener knows the global checkpoint from another shard copy (e.g., the primary), and the current shard copy is lagging behind. Today we notify the listener whenever the global checkpoint advances, regardless if it goes above the current global checkpoint known to the listener. This commit reworks this implementation. Rather than thinking of the value associated with the listener as the current global checkpoint known to the listener, we think of it as the value that the listener is waiting for the global checkpoint to advance to (inclusive). Now instead of notifying all waiting listeners when the global checkpoint advances, we only notify those that are waiting for a value not larger than the actual global checkpoint that we advanced to.

### CODE REVIEW

**Review by martijnvg — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java [hunk 1]
```diff
@@ -193,22 +190,31 @@ synchronized void globalCheckpointUpdated(final long globalCheckpoint) {
     private void notifyListeners(final long globalCheckpoint, final IndexShardClosedException e) {
         assert Thread.holdsLock(this);
         assert (globalCheckpoint == UNASSIGNED_SEQ_NO && e != null) || (globalCheckpoint >= NO_OPS_PERFORMED && e == null);
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java]: Should we break this assertion into two branches of the following if-else statement?

**Review by dnhatn — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java (hunk 1 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java]: Sure thing, I will do this in a follow-up though as I want to unblock the main effort, and make a small adjustment to another assertion in this file.

Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java]: @dnhatn See #33722.

**Pull request merged**